### PR TITLE
Handling builds with meson

### LINKED ISSUE #19: Remove numpy distutils from build
xref: https://github.com/SWxTREC/pymsis/pull/16, Python and Numpy are removing distutils support, so we should move off of that. It may be worth looking into one of the other build options (cmake, meson, scikit-build). https://numpy.org/devdocs/f2py/buildtools/index.html

### PR DESCRIPTION
**Pull request by greglucas**
This adds the ability to build with meson. Everything works locally for me on mac and Linux, but of course I don't have a Windows box and that is always the one causing failures... :(

1. can't parse `readme.rst` due to bad characters... Avoid that for now: 8d970051403abed2637ac533143e30f8572a4684
2. It seems like everything is built, but then running pytest fails because the libraries can't be found.

https://github.com/SWxTREC/pymsis/runs/8281587728?check_suite_focus=true#step:8:21
```
ImportError while importing test module 'D:\a\pymsis\pymsis\tests\test_msis.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
tests\test_msis.py:5: in <module>
    from pymsis import msis
C:\hostedtoolcache\windows\Python\3.9.13\x64\lib\site-packages\pymsis\msis.py:5: in <module>
    from pymsis import msis00f, msis20f, msis[21](https://github.com/SWxTREC/pymsis/runs/8281587728?check_suite_focus=true#step:8:22)f
E   ImportError: DLL load failed while importing msis00f: The specified module could not be found.
```

The install appears to build the wheel successfully...
https://github.com/SWxTREC/pymsis/runs/8281587728?check_suite_focus=true#step:6:1543
```
WARNING Could not tell if file was meant for purelib or platlib, so it was mapped to platlib: D:\a\pymsis\pymsis\.mesonpy-7o1ars04\install\hostedtoolcache\windows\Python\3.9.13\x64\Lib\site-packages\pymsis\msis21f.cp39-win_amd64.dll.a (C:\hostedtoolcache\windows\Python\3.9.13\x64\Lib\site-packages\pymsis\msis21f.cp39-win_amd64.dll.a)
  Copying files to wheel...
  [0/14] pymsis/msis00f.cp39-win_amd64.pyd
  [1/14] pymsis/msis00f.cp39-win_amd64.dll.a
  [2/14] pymsis/msis20f.cp39-win_amd64.pyd
  [3/14] pymsis/msis20f.cp39-win_amd64.dll.a
  [4/14] pymsis/msis21f.cp39-win_amd64.pyd
  [5/14] pymsis/msis21f.cp39-win_amd64.dll.a
  [6/14] pymsis/__init__.py
  [7/14] pymsis/msis.py
  [8/14] pymsis/msis2.0.parm
  [9/14] pymsis/msis21.parm
  [10/14] pymsis/tests/__init__.py
  [11/14] pymsis/tests/test_msis.py
  [12/14] pymsis/tests/test_regression.py
  [13/14] pymsis/tests/msis2.0_test_ref_dp.txt
  [14/14] pymsis/tests/msis2.1_test_ref_dp.txt
  Preparing metadata (pyproject.toml): finished with status 'done'
```

closes #19

### CODE REVIEW

**Comment by greglucas** [comment]
@eli-schwartz I decided to try and learn how to use meson and gave it a go. I think I'm close to getting this to build, but wondering if you have any experience with Python Windows builds and if I'm doing something wrong here. Any other advice is welcome too. (taking advantage of your previous comment ;) https://github.com/SWxTREC/pymsis/pull/16#issuecomment-1220956807 )

File: pyproject.toml [hunk 1]
```diff
@@ -1,8 +1,65 @@
 [build-system]
+build-backend = "mesonpy"
 requires = [
-    "setuptools==59.2.0",
+    "meson-python>=0.8.0",
```
Comment (rgommers) [inline pyproject.toml:4]: Can you make this `>=0.10.0`?

File: meson.build [hunk 2]
```diff
@@ -0,0 +1,75 @@
+project(
+  'pymsis',
+  'c',
+  # Note that the git commit hash cannot be added dynamically here
+  version: '0.6.0',
+  license: 'MIT',
+  meson_version: '>= 0.63.0',
```
Comment (rgommers) [inline meson.build:7]: and this `>= 0.63.3`?

Comment (greglucas) [reply meson.build]: Done

**Comment by eli-schwartz** [comment]
I'm not a Windows user myself, but I think the error message emitted by python here is saying that there are missing DLL dependencies of the module. Unfortunately, the error message doesn't say what those dependencies are.

Is it possible to use something like dumpbin or Dependency Walker on the .pyd and see what it thinks it depends on?

**Comment by rgommers** [comment]
I had a closer look. The _"Could not tell if file was meant for purelib or platlib"_ warnings are gone now, and the `.pyd` files end up in the correct location. I'm fairly sure that the missing DLL will be `libgfortran`, because that is needed for extensions from Fortran code and it's not clear to me how it will be found (`GITHUB_PATH` may not add it to the Windows library search path - although hard to be sure about that).

To build redistributable wheels, you need `delvewheel` to vendor the missing DLLs. See `cibw_before_build_win.sh` and `repair_windows.sh` at https://github.com/scipy/scipy/tree/main/tools/wheels.

I'll also point out that SciPy uses a different Mingw-w64 build, the rtools one: https://github.com/scipy/scipy/blob/main/.github/workflows/windows.yml#L42. That uses the Windows UCRT rather than the older runtime libraries, maybe that matters too.

**Comment by greglucas** [comment]
I must have overlooked that Scipy builds the libraries statically on Windows! https://github.com/scipy/scipy/blob/a082eb415192f3075921fbf3e539f2b82a53811e/scipy/meson.build#L5-L7
That is the key here and I'm able to get the Windows build working now:
https://github.com/SWxTREC/pymsis/actions/runs/3266720215
Specifically the latest commit: https://github.com/SWxTREC/pymsis/pull/21/commits/e2f0c0daef8c450985eca6272a938bcb8a8adbbb

Thank you very much for taking a look @eli-schwartz and @rgommers, your comments definitely pointed me in the right direction here! It might be good to put this into the f2py / scipy -> meson migration notes. I'll try and make some PRs for docs updates if I can think of things that would help others in the next week or so.

**Comment by rgommers** [comment]
Nice, glad that got resolved!

File: meson.build [hunk 3]
```diff
@@ -0,0 +1,89 @@
+project(
+  'pymsis',
+  'c',
+  # Note that the git commit hash cannot be added dynamically here
+  version: '0.6.0',
+  license: 'MIT',
+  meson_version: '>= 0.63.3',
```
Comment (eli-schwartz) [inline meson.build:7]: Meson 0.64.0 was released today, which means you can make a small improvement.

File: meson.build [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+  if meson.get_compiler('fortran').get_id() == 'gcc'
+    add_project_link_arguments(gcc_link_args, language: ['fortran'])
+  endif
+endif
+
+generate_f2pymod = files('tools/generate_f2pymod.py')
+
+# https://mesonbuild.com/Python-module.html
+py_mod = import('python')
+py3 = py_mod.find_installation()
```
Comment (eli-schwartz) [inline meson.build:54]:
```suggestion
py3 = py_mod.find_installation(pure: false)
```

https://mesonbuild.com/Release-notes-for-0-64-0.html#pythonfind_installation-now-accepts-pure-argument

File: pymsis/meson.build [hunk 5]
```diff
@@ -0,0 +1,9 @@
+py3.install_sources([
+  '__init__.py',
+  'msis.py',
+  'msis2.0.parm',
+  'msis21.parm',
+],
+  pure: false,
```
Comment (eli-schwartz) [inline pymsis/meson.build:7]: ... renders this obsolete, and impossible to forget in future `py3.install_sources()` calls.


```suggestion
```

**Review by eli-schwartz — CHANGES_REQUESTED** [review]

**Comment by eli-schwartz** [comment]
Aha, excellent timing on my part. :D

**Comment by greglucas** [comment]
No problem at all, @eli-schwartz! Thanks so much for commenting, I'll add that in another commit before I make a release. I've got a few more things to do before releasing a new version. The meson part of this was all pretty nice to work with, so thanks for your work on that!

**Pull request merged**